### ISSUE TITLE
Qubes Template Compatibility

### ISSUE THREAD

**Issue by bluesteal** [gramineproject/gramine#1345]
### Description of the feature

Can Gramine be implemented in a similar manner to the MirageOS firewall as described here:

https://github.com/talex5/qubes-mirage-firewall

Could it be installed via dom0 as an .rpm and created as a QubesOS template?

### Why Gramine should implement it?

It would make it more convenient to create a template that is a LibraryOS with a much smaller attack surface than the existing OSes that are provided in the QubesOS repos (Fedora, Debian).

**Comment by dimakuv** [comment]
I think you don't realize that Gramine is not a self-sufficient OS kernel?

Gramine doesn't have the FS and network stack, has only partial memory management system, doesn't have scheduling/context switching, and a bunch of other low-level primitives. Gramine relies on the host OS to provide such primitives. Thus, Gramine is not self-sufficient, it always needs some kind of kernel sitting "below" it.

**Comment by bluesteal** [comment]
What is the intended use case of Gramine?

**Comment by mkow** [comment]
> It would make it more convenient to create a template that is a LibraryOS with a much smaller attack surface than the existing OSes that are provided in the QubesOS repos (Fedora, Debian).

How do you want to make a Qubes VM template out of _any_ Library OS? The whole idea of Library OSes is to implement some system API (e.g. Linux syscall interface) using a simple API which provides the I/O and other resources. In case of Gramine we're implementing Linux syscall interface (so, we can run Linux apps) on top of PAL API (Drawbridge-like), for which we currently have Linux and Linux-SGX implementations. We also had a FreeBSD backend (allowing to run Linux binaries on FreeBSD), but it wasn't maintained and thus was removed. Another potential usecase would be to add Windows backend and run Linux apps on Windows.

So, I don't get this idea of using it in Qubes :)

**Comment by bluesteal** [comment]
Well, have you taken a look at MirageOS? It is a Unikernel that is designed to work with the QubesOS template format.
It appears to be marketing itself as a LibraryOS with a kernel that is much smaller than a typical Linux Kernel. I guess the idea
is to reduce the amount of code overall and thus reduce the attack surface.

**Comment by bluesteal** [comment]
I thought maybe Gramine would be similar to MirageOS; but its not a unikernal, it is a Linux syscall interface as you mentioned.

**Comment by dimakuv** [comment]
> I thought maybe Gramine would be similar to MirageOS; but its not a unikernal, it is a Linux syscall interface as you mentioned.

Yes, the terminology of `unikernel` vs `library OS` vs `sandbox` etc. is very confusing. It was probably a mistake to call Gramine/Graphene a Library OS in the first place? Like, Library OSes typically imply a fully-fledged kernel, just shipped in the form of libraries which can be cherry-picked for a particular application to become a self-sufficient app+kernel bundle.

But I'm not an expert in scientific terminology :) To me, a `sandbox` would probably describe Gramine better...

**Comment by mkow** [comment]
> Library OSes typically imply a fully-fledged kernel, just shipped in the form of libraries which can be cherry-picked for a particular application to become a self-sufficient app+kernel bundle.

I thought that this is the very definition of unikernels? Library OSes are rather no-I/O, usermode-runnable "emulators"?

> To me, a sandbox would probably describe Gramine better...

No, a sandbox (or at least a security sandbox) is something which isolates the host from an app. If at all, Gramine would be a reverse-sandbox (security app from the host), but it's rather the SGX which is that reverse-sandbox, Gramine is rather a compatibility layer which can be run inside SGX (+ some additional tooling around SGX).

**Comment by mkow** [comment]
> I thought maybe Gramine would be similar to MirageOS; but its not a unikernal, it is a Linux syscall interface as you mentioned.

I'm not sure why they call MirageOS a library OS, maybe their internal architecture is library-OS-like? But it's primarily a unikernel, which Gramine isn't :)

**Comment by dimakuv** [comment]
@mkow Now I'm curious how you understand Unikernels vs Library OSes vs Sandboxes.

My rough (and I guess incorrect?) understanding was like this:
- Unikernels -- application and OS kernel fused into one entity; OS kernel is minimal and supports only a single app/single process.
- Library OS -- subtype of unikernels, where the OS kernel is a set of libraries that the application must link against; the subset of libraries that the app links against may differ from case to case (e.g. an app may not want to link against the networking library).
- Sandbox -- protection measures to decrease the attack surface between two software entities. Typically protecting the host from a malicious app. Or protecting the app from a malicious host (what you called reverse-sandbox).

So to summarize, to me Gramine-SGX is a reverse security sandbox (to protect the app from the host). This then implies the design choices that Gramine makes: a lot of "shadow kernel data" state inside of Gramine to verify what the host is doing, but not everything (because e.g. TLS networking can work on the classic `listen`/`connect`/`read`/`write` sockets interface without compromising security). I guess the only design choice that is not mandated by this "reserve security sandbox" is the PAL layer that allows for compatibility with different backends.

Hm, I should go back to actual work...

**Comment by monavij** [comment]
I would go with Dmitri's definition and my understanding is that unikernels and Library OS terms are used interchangeably and they are basically single address space OSes. It's just that you can run unikernels/LibOS like Mirage in a bare metal VM environment or as a process like Gramine. In both cases you are pulling in a bunch of kernel functionality into your address space and linking directly with the application. So unikernels/Library OSes may or may not rely on a host OS depending on if they are running in  bare-metal VM or as a process on top of a kernel.
